test/_init_tests.js

```diff
@@ -207,6 +207,49 @@ describe.configure = function() {
   return new TestConfig(describe);
 };
 
+function enableYield(fn, message, runnable) {
+  if (!runnable || !runnable.constructor
+      || runnable.constructor.name !== 'GeneratorFunction') {
+    return fn(message, runnable);
+  }
+  return fn(message, done => {
+    const runner = (iterator, result) => {
```

Comment:
> Sorry, what does that mean? :-)

Promise assimilation is:

```js
Promise.resolve(1).then(value => assert(value, 1));
Promise.resolve(somePromiseThatResolvesTo1).then(value => assert(value, 1));
```

Promises are designed to be interoperable through assimilation (they do almost exactly what you do with the `value.then === function` check), but in a overspecified-spec way.

> I handled this by using a try-catch block

Done is actually not handled by the try-catch. If the generator threw, that's handled by it. But the problem is the final result value (when `iterator.done = true`) isn't waited on (it could be a promise).

>  But in your impl, there's also a try-catch, is that there for the same reason?

My try-catch and yours are equivalent, but neither handles the `done` callback (see above).

> not sure how it is less than the done impl.

Multiple closures are used when the yielded value is a promise (the then blocks). This one's really a nit-pick.

> Also, I still think there is value to support setTimeout

https://github.com/ampproject/amphtml/pull/9601#issuecomment-305370522